daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/BuildDisclosureTableTest.scala

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.data.ImmArray
+import com.daml.lf.interpretation.Error.DisclosurePreprocessing
+import org.scalatest.Inside
+import org.scalatest.freespec.AnyFreeSpec
+import org.scalatest.matchers.should.Matchers
+
+class BuildDisclosureTableTest extends AnyFreeSpec with Inside with Matchers {
+
+  import BuildDisclosureTableTest._
+  import ExplicitDisclosureLib._
+
+  "buildDisclosureTable" - {
+    "disclosure preprocessing" - {
+      "template does not exist" in {
+        val error = intercept[SError.SErrorDamlException] {
+          Speedy.buildDiscTable(ImmArray(disclosedHouseContractInvalidTemplate), pkg.pkgInterface)
+        }
+
+        error shouldBe SError.SErrorDamlException(
+          DisclosurePreprocessing(
+            DisclosurePreprocessing.NonExistentTemplate(
+              disclosedHouseContractInvalidTemplate.templateId
+            )
+          )
+        )
+      }
+
+      "disclosed contract key has no hash" in {
+        val error = intercept[SError.SErrorDamlException] {
+          Speedy.buildDiscTable(ImmArray(disclosedHouseContractNoHash), pkg.pkgInterface)
+        }
+
+        error shouldBe SError.SErrorDamlException(
+          DisclosurePreprocessing(
+            DisclosurePreprocessing.NonExistentDisclosedContractKeyHash(
+              contractId,
+              disclosedHouseContractNoHash.templateId,
+            )
+          )
+        )
+      }
+
+      "duplicate disclosed contract IDs" in {
+        val error = intercept[SError.SErrorDamlException] {
+          Speedy.buildDiscTable(
+            ImmArray(disclosedHouseContract, disclosedHouseContract),
+            pkg.pkgInterface,
+          )
+        }
+
+        error shouldBe SError.SErrorDamlException(
+          DisclosurePreprocessing(
+            DisclosurePreprocessing.DuplicateContractIds(houseTemplateId)
+          )
+        )
+      }
+
+      "duplicate disclosed contract key hashes" in {
+        val error = intercept[SError.SErrorDamlException] {
+          Speedy.buildDiscTable(
+            ImmArray(disclosedHouseContract, disclosedHouseContractWithDuplicateKey),
+            pkg.pkgInterface,
+          )
+        }
+
+        error shouldBe SError.SErrorDamlException(
+          DisclosurePreprocessing(
+            DisclosurePreprocessing.DuplicateContractKeys(houseTemplateId)
```

Comment:
Slightly out of the scope for this PR but why does this error not include the key hash?

Reply:
Ah, oversight by myself. Corrected this on PR https://github.com/digital-asset/daml/pull/14592